### ISSUE TITLE
Allow header field `Cache-Control` in preflight requests

### ISSUE THREAD

**Issue by rgbkrk** [jupyterhub/binderhub#315]
Hopefully I can just code around this. However, I'm running into an issue with the `Access-Control-Allow-Headers` not including `Cache-Control`:

```
Failed to load https://mybinder.org/build/gh/nteract/vdom/master: Request header field Cache-Control is not allowed by Access-Control-Allow-Headers in preflight response
```

Full request structure (from Chrome dev tools):

```
:authority:mybinder.org
:method:OPTIONS
:path:/build/gh/nteract/vdom/master
:scheme:https
accept:*/*
accept-encoding:gzip, deflate, br
accept-language:en-US,en;q=0.9
access-control-request-headers:cache-control
access-control-request-method:GET
cache-control:no-cache
origin:http://127.0.0.1:3000
pragma:no-cache
referer:http://127.0.0.1:3000/
user-agent:Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36
```

**Comment by yuvipanda** [comment]
What's the action item here? Am very behind on CORS changes, so not sure what needs to be done on the binderhub side :)

**Comment by rgbkrk** [comment]
Binderhub would need to return `Cache-Control` in its [`Access-Control-Allow-Headers`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers) response Header. Sadly a `*` value does not work as it isn't implemented in all browsers.

**Comment by rgbkrk** [comment]
The other easy route is to just stick everything from the client's `Access-Control-Request-Headers`  into the `Access-Control-Allow-Headers`.

By the way, I ended up working around this by getting the ajax lib we were using to not even send `Access-Control-Request-Headers`.

**Comment by yuvipanda** [comment]
Hmm, I wonder if this is just Tornado's default options handling, and we need to overwrite it. I don't see us doing anything at all about OPTIONS atm.

**Comment by willingc** [comment]
@yuvipanda I think you are correct about Tornado and setting default headers for a request.

https://github.com/jupyterhub/binderhub/blob/master/binderhub/base.py#L12

http://www.tornadoweb.org/en/stable/web.html#tornado.web.RequestHandler.set_default_headers

**Comment by minrk** [comment]
It's not tornado blocking the request, it's the browser blocking the CORS request because the ajax lib requested a header and tornado didn't explicitly allow it with `Access-Control-Allow-Headers`, which if left unspecified is assumed to be very strict. Adding a more liberal `Access-Control-Allow-Headers` to set_default_headers sould do it, as @willingc suggested.

**Comment by Miniland1333** [comment]
Hello! Do you guys have an update on this issue? I know it has been six months since the last comment but I am still getting this issue. I believe that the additional Access-Control-Request-Header: Cache-Control is from a service-worker since disabling the service-worker temporarily remediates the issue! Please let me know if you come up with anything since I would like to keep the service-worker enabled.


![image](https://user-images.githubusercontent.com/12994749/43982043-2eab60ae-9ca9-11e8-9f4a-b8be36d44e28.png)

**Comment by betatim** [comment]
I don't think anyone has worked on this. Would make a good first contribution! Thanks for your help.

**Comment by Miniland1333** [comment]
I just made a pull request after testing. Just adding access-control-allow-headers: "cache-control" allows the initial OPTIONS to succeed and therefore also allows the eventsource to proceed as normal. Thank you @willingc for finding the exact spot for me to add the line of code!


Based on my experience, I will also suggest additions to the installation portion of the contributing guide since there were a few components not explicitly listed (python3-pip, npm, curl, socat, docker). Most developers have these installed already, but when working from a fresh ubuntu installation (upgrade to 18.04 LTS) it was a hindrance to figure out these unlisted dependencies.


If the pull request is accepted I'll update this thread if my production site ceases to have this error when communicating with mybinder through the serviceworker. Thanks for all your help so far!

**Comment by betatim** [comment]
It would be great to have a link to the site where you are using this. Also the library/code you are using to talk to binderhub. There are a few out there already that do this but mostly hard to find. So starting a list would be good.

👍 to listing what other tools you need to have in `CONTRIBUTING.md`. I wouldn't go as far as giving specific instructions because we can't cover Ubuntu, OSX, Arch linux, ... times the five different ways you could install them on each of these systems. I'd instead list them and say "we assume you have your favourite way of installing/providing these tools".

**Comment by Miniland1333** [comment]
I'm using @minrk 's [Thebelab](https://github.com/minrk/thebelab) project to embed into an HTML webpage. I knew of Thebe but didn't know Thebelab had resurrected the project. I wish Thebelab had higher visibility since it is perfect for external websites to communicate with Binderhub!

An example of Thebelab that I'm working on is [Chuck Serverance's e-textbook](https://eng.libretexts.org/Textbook_Maps/Computer_Science/Map%3A_Python_for_Everybody_(Severance)/4%3A_Functions/4.04%3A_Random_numbers). (You many need to refresh the page once due to the above issue)

![image](https://user-images.githubusercontent.com/12994749/44311073-a8aa7c80-a395-11e8-9f38-e987eb21be50.png)

**Comment by betatim** [comment]
Nice work! Worked first time for me on Firefox.

I would recommend you change the setup so that it only requests a binder instance when a user performs an explicit action like clicking "Run" or "start the kernel". (For example the (somewhat hard to see) "Make it live" button on https://sebasguts.github.io/thebelab_test_gap/chap42). 

If you get a lot of traffic it will spawn one binder per visitor which means you will run into the limit of 100 concurrent binders per repository. We would probably ban you till you fix it if that happened.

Didn't know thebelab used service workers.

**Comment by Miniland1333** [comment]
I've removed the requestKernel: true, which means that binder instances are now manually requested only. I might integrate thebelab into my analytics to gauge how popular it is.

Thebelab itself does not use serviceworkers but the site as a whole does. I'm glad we could deal with the issue binderhub-side since the serviceworkers are bundled as part of the PaaS.

**Comment by betatim** [comment]
Fixed in #641
